### ISSUE TITLE
Would you mind adding a driver for the 2016+ MacBook Pros (Links)

### ISSUE THREAD

**Issue by DiegoMagdaleno** [sebanc/brunch#258]
Hi, I am a ChromeOS enthusiast, and I really like your work, I use it on my MacBook Pro, and most of it works, except for the keyboard (sound doesn't but I don't know if you would bother adding that too), kernel modules have been created to run this, and I would like to see if you would bother adding them, I tried on my own but if I am honest I didn't get anywhere.
 
Thank you for your amazing work :).

Here are the links:

- [Keyboard and Touchpad Driver](https://github.com/roadrunner2/macbook12-spi-driver)

And if you wish:

- [Sound driver](https://github.com/davidjo/snd_hda_macbookpro)

Thanks a lot again. Greetings!

**Comment by sebanc** [comment]
Hi,

I added the apple spi keyboard and touchpad driver in the unstable build available here:
https://github.com/sebanc/brunch-testing/releases

Could you try it ?

For sound, the state of the module is too experimental and I don't want to risk breaking people's speakers... However, you can build the module directly in brunch if you want to add it:
- Download chromebrew
- Clone the sound driver git
- Run make
- Replace the module with the same name in "/lib/modules/4.19..../kernel/sound/...." with your built one.

**Comment by DiegoMagdaleno** [comment]
Thanks a lot, and sure I will do, thanks a lot for your work :), I report back if the sound is working in a bit!

**Comment by DiegoMagdaleno** [comment]
Okay so I downloaded the new one, i just have to wait now for the download of the ChromeOS recovery, do the process and then I notify you.

**Comment by DiegoMagdaleno** [comment]
There is only 1 issue:

- Moving with the trackpad doesnt work, clicking does but you cant move the cursor, also two finger right click doesnt work, in Linux it does so I don't know if this is an issue with the driver or chromeOS

**Comment by sebanc** [comment]
Try with "options=alt_touchpad_config"

**Comment by DiegoMagdaleno** [comment]
I try the touchpad options later, for now, sound isnt working with the driver, i dont know if I did this the right way or not.

![ls of /lib/modules](https://user-images.githubusercontent.com/38844659/83791123-e8bc1b80-a65e-11ea-8653-8571443eefb5.png)

**Comment by sebanc** [comment]
No, actually I did not provide the complete path. Run "find /lib/modules | grep "cirrus" to identify the path of the original module and then replace the existing file with yours.

**Comment by DiegoMagdaleno** [comment]
Okay so I can finally report back and here are the results:

- Trackpad only works for clicking you cant move with the trackpad 
- Sound isnt working

The rest is working flawlessly and the work around for hibernation on Linux works on ChromeOS too, i trust that in the future support for this MacBooks will be getting better in Linux, if you don't mind I would find very very useful to have a git history when you added the MacBook pro keyboard driver, so in the future I can contribute on my own, thanks a lot for your amazing work, greetings!

**Comment by sebanc** [comment]
Could you attach a dmesg ? (run "dmesg > ~/Downloads/dmesg.txt" in crosh shell and attach dmesg.txt from your downloads folder).

**Comment by DiegoMagdaleno** [comment]
Sure here you go
[dmesg.txt](https://github.com/sebanc/brunch/files/4732091/dmesg.txt)

**Comment by sebanc** [comment]
The sound module does not seem to be probed, could you try to reboot, then press CTRL+ALT+F2, login as "root" and run "insmod <path_to_the_sound_module>", then CTRL+ALT+F1, login, test sound and attach another dmesg.

**Comment by DiegoMagdaleno** [comment]
Looks like it can’t load the module because it contains an unknown symbol 

PS: Sorry for the Photo in my iPad I only have GitHub there

![image](https://user-images.githubusercontent.com/38844659/83803952-01363100-a673-11ea-8302-a3b361425d4d.jpeg)

**Comment by sebanc** [comment]
Thanks for the photo, could you build the module again and post the output of the "make" command here ? I think something went wrong at build time.

**Comment by DiegoMagdaleno** [comment]
No problem and yes sure, I could, and also I ran scrip+evtest in the Apple SPI Touchpad here is it, hope some of it can helps, it does detect when it is being touched. And sure I show the modules in a bit. 

And here is the evtest on the touchpad:
[chromeOSOnMacbookDebugTouchpad.txt](https://github.com/sebanc/brunch/files/4732449/chromeOSOnMacbookDebugTouchpad.txt)

**Comment by DiegoMagdaleno** [comment]
And here is the output of the sound driver while being built. 
[buildingCirrusDriverIssue.txt](https://github.com/sebanc/brunch/files/4732454/buildingCirrusDriverIssue.txt)

**Comment by DiegoMagdaleno** [comment]
The Apple Magic Trackpad works and i might have an idea of modifying the configuration.

[Got the idea from here ](https://neverware.zendesk.com/hc/en-us/community/posts/360000764527-Neverware-MacBook-Air-Touchpad-Driver)

I report back in a bit.

**Comment by DiegoMagdaleno** [comment]
Okay so I reconfigured the trackpad and that didnt see to help, the trackpad is getting detected however and tapping on it does register an event I think, probably is some configuration issue on ChromeOS?

**Comment by DiegoMagdaleno** [comment]
@sebanc So a little update on some stuff:

- Touchpad isnt working, however it is properly detected, this might be a configuration issue of ChromeOS itself.

- Sound undefined symbol is because this is a GNULIBC exclusive symbol, what does this mean? This means only GNU operating systems can use this, and ofc we know ChromeOS doesnt use GNULIBC, however a quick fix is this code:

```c
unsigned long __stack_chk_guard;
void __stack_chk_guard_setup(void)
{
     __stack_chk_guard = 0xBAAAAAAD;
}

void __stack_chk_fail(void)                         
{
 /* Error message */                                 
}
```
insmod doesnt fail anymore, however there is still no sound, this might be a configuration issue of ChromeOS at the end of the day.

**Comment by sebanc** [comment]
I think the correct way of fixing the "stack_chk_guard" issue would be to add
`CFLAGS_snd-hda-codec-cirrus.o := -fno-stack-protector`

after the below lines in the Makefile:
```
snd-hda-codec-cirrus-objs :=	patch_cirrus.o
obj-$(CONFIG_SND_HDA_CODEC_CIRRUS) += snd-hda-codec-cirrus.o
```

Could try this build method, replace the kernel module and post another dmesg ?

**Comment by sebanc** [comment]
I might have broken something related to sound in the previous brunch testing release, could you try to build the sound module with the above fix using the new test release:
https://github.com/sebanc/brunch-testing/releases

**Comment by DiegoMagdaleno** [comment]
Yeah I will test this today and notify any progress, sorry for not replying before GitHub wasn’t showing the notification

**Comment by sebanc** [comment]
@canayaf713 You should now have access to the testing repo. This feature will also be included in the next standard release which should happen during the weekend.

**Comment by canayaf713** [comment]
Thanks! My keyboard is working now, however I am having the same trackpad issue. I will try to figure out what's causing the problem and update if there is any progress.

**Comment by DiegoMagdaleno** [comment]
@canayaf713 for your trackpad could you try https://neverware.zendesk.com/hc/en-us/community/posts/360000764527-Neverware-MacBook-Air-Touchpad-Driver 

Trying something like that, click works, but not scrolling for reason.

**Comment by DiegoMagdaleno** [comment]
A possible fix will be to use libinput, since that works on Linux, but I dont know if this is possible, I know cloudready does that.

**Comment by sebanc** [comment]
Unfortunately, cmt is the equivalent for libinput in chromeos and I don't think I can replace it.

**Comment by canayaf713** [comment]
@DiegoMagdaleno Does your trackpad work? I tried the link you provided but only scrolling and clicking works.

**Comment by DiegoMagdaleno** [comment]
@canayaf713 it doesnt however, "Scrolling" is more progress than I ever got, continue playing with the configuration and you might get it.

**Comment by Joshagos** [comment]
WiFi still isn't working on my 2017 MacBook Air.
